### PR TITLE
Add availableDriver search, improve ride select queries

### LINKED ISSUE #24: Driver availability Query
In this data model the most complex query there is is the one to figure out what drivers are available to drive at a given time. 

You will get 5 variables coming from php `$dayOfTheWeek` `$timeOfDayStart` `$timeOfDayEnd` `$datetimeStart` and `$datetimeEnd`

This query must 
- find all users who are drivers
- Join to Driver Availability and see `$dayOfTheWeek` is in the `days` set between `$timeOfDayStart` and `$timeOfDayEnd`
- Join to Driver AvailabilityException and make sure they don't have any exceptions which intersect with `$datetimeStart` and `$datetimeEnd`
- They must not be scheduled to drive any other ride during that time. 

image

### LINKED ISSUE #52: Add Populate argument to GET queries.
We should add the ability to shallow populate relations. Right now I really only care about doing this from the rides table. But a system to do it from any query in the future might be nice. 

# /api/rides.php
Should return something like this
```json
[{
	"id": 1,
	"passengerID": 4,
        "driverID": 5,
	"apptStart": "2018-04-23 09:00:00",
	"apptEnd": "2018-04-23 10:00:00",
	"numMiles": null,
	"totalMinutes": null,
	"pickupTime": "2018-04-23 08:30:00",
	"wheelchairVan": 0,
	"status": "Scheduled",
	"pickupStreetAddress": "855 publishers parkway",
	"pickupCity": "Webster",
	"apptStreetAddress": "45 Webster Commons Blvd #201",
	"apptCity": "Webster",
	"created": "2018-02-01 08:00:00",
	"modified": "2018-02-05 14:51:08"
}]
```

# /api/rides.php?populate=passenger
Should return something like this

```json
[{
	"id": 1,
	"passenger": {
	    "id": 4,
	    "firstName": "Main",
	    "lastName": "Passenger",
	    "role": "passenger",
	    "phone": "2435254235",
	    "email": "passanger@websterwasps.com",
	    "registered": "2018-01-01 00:00:00",
	    "lastLogin": "2018-03-01 00:00:00",
	    "wantsSMS": 1,
	    "wantsEmail": 1
         },
        "driverID": 5,
	"apptStart": "2018-04-23 09:00:00",
	"apptEnd": "2018-04-23 10:00:00",
	"numMiles": null,
	"totalMinutes": null,
	"pickupTime": "2018-04-23 08:30:00",
	"wheelchairVan": 0,
	"status": "Scheduled",
	"pickupStreetAddress": "855 publishers parkway",
	"pickupCity": "Webster",
	"apptStreetAddress": "45 Webster Commons Blvd #201",
	"apptCity": "Webster",
	"created": "2018-02-01 08:00:00",
	"modified": "2018-02-05 14:51:08"
}]
```

# /api/rides.php?populate=driver,passenger
Should return something like this

```json
[{
	"id": 1,
	"passenger": {
	    "id": 4,
	    "firstName": "Main",
	    "lastName": "Passenger",
	    "role": "passenger",
	    "phone": "2435254235",
	    "email": "passanger@websterwasps.com",
	    "registered": "2018-01-01 00:00:00",
	    "lastLogin": "2018-03-01 00:00:00",
	    "wantsSMS": 1,
	    "wantsEmail": 1
         },
        "driver": {
	    "id": 5,
	    "firstName": "Weekend",
	    "lastName": "Driver",
	    "role": "driver",
	    "phone": "5852167818",
	    "email": "weekendDriver@websterwasps.com",
	    "registered": "2018-01-01 00:00:00",
	    "lastLogin": "2018-03-01 00:00:00",
	    "wantsSMS": 1,
	    "wantsEmail": 1
        },
	"apptStart": "2018-04-23 09:00:00",
	"apptEnd": "2018-04-23 10:00:00",
	"numMiles": null,
	"totalMinutes": null,
	"pickupTime": "2018-04-23 08:30:00",
	"wheelchairVan": 0,
	"status": "Scheduled",
	"pickupStreetAddress": "855 publishers parkway",
	"pickupCity": "Webster",
	"apptStreetAddress": "45 Webster Commons Blvd #201",
	"apptCity": "Webster",
	"created": "2018-02-01 08:00:00",
	"modified": "2018-02-05 14:51:08"
}]
```

### PR DESCRIPTION
**Pull request by lzganec**
Fix #52, Fix #24 

@BLamy 
1. The populate argument works fine but it puts the new fields at the bottom of the array since I've "renamed" them. If you don't want the json field to read "driver" instead of "driverID" then the order will remain as it was.
2. For the find driver query I've made the new search.php endpoint. We can discuss it if we want to organize it a bit different, but I think a new endpoint is clearer than stacking everything on the users.php, although logically it belongs to users.php in a way.

### CODE REVIEW

File: server/dao/RideDAO.php [hunk 1]
```diff
@@ -35,7 +43,14 @@ public function getRides($page = 0, $numberPerPage = 10)
             $stmt->execute();
             $stmt->setFetchMode(PDO::FETCH_CLASS, "Ride");
             while ($ride = $stmt->fetch()) {
-                $data[] = $ride->getRideInfo();
+                $ride = $ride->getRideInfo();
+
+                if ($populate)
+                {
+                   $this->populateIDs($ride);
```
Comment (BLamy) [inline server/dao/RideDAO.php]: Will work for now but we will probably want to switch this to a join for production.

File: server/dao/UserDAO.php [hunk 2]
```diff
@@ -42,6 +42,36 @@ function findByName($name) {
...
     }
 
+    function findAvailableDrivers($dayOfTheWeek, $timeOfDayStart, $timeOfDayEnd, $datetimeStart, $datetimeEnd)
+    {
+       try {
+          $query = "SELECT user.id, user.firstName, user.lastName, user.phone, user.email FROM user 
+            LEFT JOIN (availability) ON (user.id = availability.driverID) 
+            LEFT JOIN (availabilityexclusion) ON (user.id = availabilityexclusion.driverID)
+            WHERE user.role = 'driver'
+              AND FIND_IN_SET(:dayOfTheWeek, availability.days)>0
```
Comment (BLamy) [inline server/dao/UserDAO.php]: 👍

File: server/api/search.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<?php
+header('Content-Type: application/json');
+
+include '../env.php';
+require_once "../db.class.php";
+$db = new DB($host, $port, $name, $user, $pass); // From dbinfo.php
+
+switch ($_SERVER['REQUEST_METHOD'])
+{
+   case "GET":
```
Comment (BLamy) [inline server/api/search.php]: I was thinking that search.php would be more for the search bar at the top of the application. I think this makes more sense in drivers.php

Comment (lzganec) [reply server/api/search.php]: Yeah, sounds good. If you'll need to use the function right away you can move it immediately. If not, I'll move it before the next pull req.

Comment (BLamy) [reply server/api/search.php]: Don't worry about it I'm about to do another iteration of hooking up the frontend to the api so I'll take care of it.

**Pull request merged**